### ISSUE TITLE
Upgrade to Svelte 4, fix accessibility and linting warnings

### ISSUE THREAD

**Issue by atjn** [josdejong/svelte-jsoneditor#296]
Hello and thank you for making this awesome component!

When bundling it in my project, I receive 23 a11y warnings:
<details>

<summary>Click to see all warnings</summary>

```
transforming (76) node_modules/esm-env/prod-browser.js18.48.12 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/plugins/value/components/BooleanToggle.svelte:27:0 A11y: <div> with mousedown handler must have an ARIA role
25: </script>
26: 
27: <div
    ^
28:   class="jse-boolean-toggle"
29:   class:jse-readonly={readOnly}
18.48.12 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/plugins/value/components/ReadonlyValue.svelte:32:0 A11y: <div> with click, dblclick handlers must have an ARIA role
30: 
31: <!-- svelte-ignore a11y-click-events-have-key-events -->
32: <div
    ^
33:   data-type="selectable-value"
34:   class={getValueClass(value, parser)}
transforming (142) node_modules/svelte-jsoneditor/utils/uniqueId.js18.48.13 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/controls/EditableDiv.svelte:109:0 A11y: <div> with keydown handler must have an ARIA role
107: </script>
108: 
109: <div
     ^
110:   class={classnames('jse-editable-div', valueClass, { 'jse-short-text': shortText })}
111:   contenteditable="true"
transforming (540) node_modules/svelte-jsoneditor/components/modals/transformModalState.js18.48.14 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modals/popup/AbsolutePopupEntry.svelte:83:0 A11y: <div> with mousedown, keydown handlers must have an ARIA role
81: />
82: 
83: <div
    ^
84:   bind:this={refRootPopup}
85:   class="jse-absolute-popup"
18.48.15 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/TreeMode.svelte:1595:0 A11y: <div> with keydown, mousedown, contextmenu handlers must have an ARIA role
1593: <svelte:window on:mousedown={handleWindowMouseDown} />
1594: 
1595: <div
      ^
1596:   class="jse-tree-mode"
1597:   class:no-main-menu={!mainMenuBar}
18.48.15 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/tablemode/TableMode.svelte:1291:0 A11y: <div> with mousedown, keydown, contextmenu handlers must have an ARIA role
1289: </script>
1290: 
1291: <div
      ^
1292:   class="jse-table-mode"
1293:   class:no-main-menu={!mainMenuBar}
transforming (1073) node_modules/svelte-jsoneditor/logic/validation.js18.48.16 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/controls/Message.svelte:23:2 A11y: <div> with click handler must have an ARIA role
21: <div class="jse-message jse-{type}">
22:   <!-- svelte-ignore a11y-click-events-have-key-events -->
23:   <div class="jse-text" class:jse-clickable={!!onClick} on:click={handleClick}>
      ^
24:     <div class="jse-text-centered">
25:       {#if icon}
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:491:8 A11y: <div> with click handler must have an ARIA role
489:       {#if expanded}
490:         <!-- svelte-ignore a11y-click-events-have-key-events -->
491:         <div
             ^
492:           class="jse-insert-selection-area jse-inside"
493:           data-type="insert-selection-area-inside"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:498:8 A11y: <div> with click handler must have an ARIA role
496:       {:else}
497:         <!-- svelte-ignore a11y-click-events-have-key-events -->
498:         <div
             ^
499:           class="jse-insert-selection-area jse-after"
500:           data-type="insert-selection-area-after"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:560:10 A11y: <div> with click handler must have an ARIA role
558:         {#if !root}
559:           <!-- svelte-ignore a11y-click-events-have-key-events -->
560:           <div
               ^
561:             class="jse-insert-selection-area jse-after"
562:             data-type="insert-selection-area-after"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:612:8 A11y: <div> with click handler must have an ARIA role
610:       {#if expanded}
611:         <!-- svelte-ignore a11y-click-events-have-key-events -->
612:         <div
             ^
613:           class="jse-insert-selection-area jse-inside"
614:           data-type="insert-selection-area-inside"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:619:8 A11y: <div> with click handler must have an ARIA role
617:       {:else if !root}
618:         <!-- svelte-ignore a11y-click-events-have-key-events -->
619:         <div
             ^
620:           class="jse-insert-selection-area jse-after"
621:           data-type="insert-selection-area-after"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:676:10 A11y: <div> with click handler must have an ARIA role
674:         {#if !root}
675:           <!-- svelte-ignore a11y-click-events-have-key-events -->
676:           <div
               ^
677:             class="jse-insert-selection-area jse-after"
678:             data-type="insert-selection-area-after"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:710:8 A11y: <div> with click handler must have an ARIA role
708:       {#if !root}
709:         <!-- svelte-ignore a11y-click-events-have-key-events -->
710:         <div
             ^
711:           class="jse-insert-selection-area jse-after"
712:           data-type="insert-selection-area-after"
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONNode.svelte:422:0 A11y: <div> with mousedown, mousemove, mouseover, mouseout handlers must have an ARIA role
420: </script>
421: 
422: <div
     ^
423:   class={classnames(
424:     'jse-json-node',
18.48.17 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/menu/SearchBox.svelte:101:4 A11y: Non-interactive element <form> should not be assigned mouse or keyboard event listeners.
 99: {#if show}
100:   <div class="jse-search-box">
101:     <form class="jse-search-form" on:submit={handleSubmit} on:keydown={handleKeyDown}>
         ^
102:       {#if !readOnly}
103:         <button
transforming (1089) node_modules/svelte-jsoneditor/components/controls/JSONPreview.svelte18.48.18 [vite-plugin-svelte] /node_modules/svelte-select/Select.svelte:694:20 A11y: <div> with mouseover, click, keydown handlers must have an ARIA role
692:             {:else if filteredItems.length > 0}
693:                 {#each filteredItems as item, i}
694:                     <div
                         ^
695:                         on:mouseover={() => handleHover(i)}
696:                         on:focus={() => handleHover(i)}
18.48.18 [vite-plugin-svelte] /node_modules/svelte-select/Select.svelte:743:20 A11y: <div> with click, keydown handlers must have an ARIA role
741:             {#if multiple}
742:                 {#each value as item, i}
743:                     <div
                         ^
744:                         class="multi-item"
745:                         class:active={activeValue === i}
18.48.18 [vite-plugin-svelte] /node_modules/svelte-select/Select.svelte:669:0 A11y: <div> with mousedown handler must have an ARIA role
667: <svelte:window on:click={handleClickOutside} on:keydown={handleKeyDown} />
668: 
669: <div
     ^
670:     class="svelte-select {containerClasses}"
671:     class:multi={multiple}
transforming (1345) node_modules/lodash-es/_createHybrid.js18.48.19 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/controls/contextmenu/ContextMenu.svelte:53:0 A11y: <div> with keydown handler must have an ARIA role
51: </script>
52: 
53: <div class="jse-contextmenu" bind:this={refContextMenu} on:keydown={handleKeyDown}>
    ^
54:   {#each items as item}
55:     {#if isMenuButton(item)}
transforming (1359) node_modules/svelte-jsoneditor/utils/jsonPointer.js18.48.19 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/CollapsedItems.svelte:31:0 A11y: <div> with mousemove handler must have an ARIA role
29: </script>
30: 
31: <div
    ^
32:   class="jse-collapsed-items"
33:   class:jse-selected={selected}
18.48.19 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/modes/treemode/JSONKey.svelte:60:2 A11y: <div> with dblclick handler must have an ARIA role
58:   />
59: {:else}
60:   <div data-type="selectable-key" class={getKeyClass(key)} on:dblclick={handleKeyDoubleClick}>
      ^
61:     {#if searchResultItems}
62:       <SearchResultHighlighter text={context.normalization.escapeValue(key)} {searchResultItems} />
transforming (1539) node_modules/lodash-es/_metaMap.js18.48.20 [vite-plugin-svelte] /node_modules/svelte-jsoneditor/components/controls/DropdownButton.svelte:38:0 A11y: <div> with click handler must have an ARIA role
36: 
37: <!-- svelte-ignore a11y-click-events-have-key-events -->
38: <div class="jse-dropdown-button" {title} on:click={handleClick}>
    ^
39:   <slot name="defaultItem" />
```

</details>

I try my best to make my apps accessible, and this amount of warnings makes me think that there are some issues in `jsoneditor`. I also would prefer it if my build output could be less filled with these warnings, as it is hard to find warnings for my own code now.

Is this something you are aware of? Is there an explanation for the warnings? I might be able to help resolve them if you want me to :)

**Comment by josdejong** [comment]
Thanks for reporting Anton. Those warnings are new since Svelte 4, which came out a couple of weeks ago I think. Upgrading to Svelte 4 requires some work. I made a start with it a few days ago but hit some issues like eslint breaking and requiring changing the config, and the generation of the TS types breaking. It wil require some effort to figure it all out and not all of the issues are under my control.

**Comment by atjn** [comment]
That makes sense, I was not aware the update added more warnings. Thanks for your work, I will look forward to the Svelte 4 update :)

**Comment by josdejong** [comment]
I'm working on the Svelte 4 upgrade in the following branch: https://github.com/josdejong/svelte-jsoneditor/tree/chore/svelte4

Most issues and warnings are fixed, and the biggest open issue is the following: https://github.com/Swatinem/rollup-plugin-dts/issues/162#issuecomment-1677556479. If anyone knows a solution please let me know.

**Comment by josdejong** [comment]
I think I've everything sorted now, see PR #303 (not yet merged).

**Comment by josdejong** [comment]
Published now in `v0.18.0`
